core/internal/filetransfer/file_transfer_default.go

```diff
@@ -87,28 +87,39 @@ func (ft *DefaultFileTransfer) Upload(task *Task) error {
 		task.Size = stat.Size() - task.Offset
 	}
 
-	reader := io.NewSectionReader(file, task.Offset, task.Size)
-
-	progressReader, err := NewProgressReader(
-		reader,
-		task.Size,
-		func(processed int, total int) {
-			if task.ProgressCallback != nil {
-				task.ProgressCallback(processed, total)
-			}
-
-			ft.fileTransferStats.UpdateUploadStats(FileUploadInfo{
-				FileKind:      task.FileKind,
-				Path:          task.Path,
-				UploadedBytes: int64(processed),
-				TotalBytes:    int64(total),
-			})
-		},
-	)
-	if err != nil {
-		return err
+	// Due to historical mistakes, net/http interprets a 0 value of
+	// Request.ContentLength as "unknown" if the body is non-nil, and
+	// doesn't send the Content-Length header which is usually required.
+	//
+	// To have it understand 0 as 0, the body must be set to nil or
+	// the NoBody sentinel.
+	var requestBody any
+	if task.Size == 0 {
+		requestBody = http.NoBody
```

Comment:
THIS! Thank you so much for figuring this out, I regret that I didn't manage to but am so glad you were able to.